src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,1101 @@
+            return;
+        }
+
+        $tournament = $this->load_tournament($tournamentId);
+        $this->save_tournament($tournament);
+    }
+
+    /**
+     * Retrieve the status of a tournament
+     *
+     * @param BMTournament $tournament
+     * @return string
+     */
+    protected function get_tournament_status(BMTournament $tournament) {
+        if (BMTournamentState::END_TOURNAMENT == $tournament->tournamentState) {
+            $status = 'COMPLETE';
+        } elseif (BMTournamentState::CANCELLED == $tournament->tournamentState) {
+            $status = 'CANCELLED';
+        } elseif (BMTournamentState::JOIN_TOURNAMENT == $tournament->tournamentState) {
+            $status = 'OPEN';
+        } else {
+            $status = 'ACTIVE';
+        }
+
+        return $status;
+    }
+
+    /**
+     * Perform a user action on a tournament
+     *
+     * @param int $userId
+     * @param int $tournamentId
+     * @param string $action
+     * @param array $buttonNameArray
+     * @return bool
+     */
+    public function act_on_tournament($userId, $tournamentId, $action, $buttonNameArray = NULL) {
+        switch ($action) {
+            case 'join':
+                return $this->join_tournament($userId, $tournamentId, $buttonNameArray);
+            case 'leave':
+                return $this->leave_tournament($userId, $tournamentId);
+            case 'cancel':
+                return $this->cancel_tournament($userId, $tournamentId);
+            case 'changeButton':
+                return $this->change_button_in_tournament($userId, $tournamentId, $buttonNameArray);
+            default:
+                $this->set_message('Invalid action on tournament');
+                return NULL;
+        }
+    }
+
+    /**
+     * Attempt to add a player to a tournament
+     *
+     * @param int $userId
+     * @param int $tournamentId
+     * @param array $buttonNameArray
+     * @return bool
+     */
+    protected function join_tournament($userId, $tournamentId, $buttonNameArray) {
+        try {
+            $tournament = $this->load_tournament($tournamentId);
+
+            // convert button names to button IDs
+            $buttonIdArray = $this->game()->retrieve_button_ids(
+                array_fill(0, count($buttonNameArray), $userId),
+                $buttonNameArray
+            );
+
+            if (!$this->validate_join_tournament($userId, $tournamentId, $tournament, $buttonIdArray)) {
+                return NULL;
+            }
+
+            $this->resolve_random_button_selection_tournament($buttonIdArray);
+
+            $success = $this->add_user_to_tournament($userId, $tournamentId, $tournament->nPlayers, $buttonIdArray);
+
+            if (!$success) {
+                // something's gone wrong between validation and attempting to add the user
+                $tournament = $this->load_tournament($tournamentId);
+                $validation = $this->validate_join_tournament($userId, $tournamentId, $tournament);
```

Comment:
This line shouldn't be reachable unless something has gone wrong, but nonetheless it is malformed --- `validate_join_tournament()` takes four arguments, and you only have 3 here.

Reply:
Argument added